<h1><center>Segmenting and Clustering Neighborhoods in Toronto</center></h1>


Author: Arturk Mammadli

This notebook includes Questions 1, 2 & 3 of the Peer-graded Assignment: Segmenting and Clustering Neighborhoods in Toronto.

# Question 1

### Importing Data

In [1]:
import pandas as pd
import requests

In [2]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

In [3]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [4]:
len(wiki_data), type(wiki_data)

(3, list)

In [5]:
wiki_data = wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

In [6]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [7]:
df = df.groupby(['Postal Code']).head()
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [8]:
df.Neighbourhood.str.count("Not assigned").sum()

0

In [9]:
df = df.reset_index()
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

In [10]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [11]:
df.shape

(103, 3)

Answer to Question 1: We have 103 rows and 3 columns.

# Question 2

### Installing Geocoder

In [12]:
pip install geocoder

     |████████████████████████████████| 98 kB 5.7 MB/s eta 0:00:011
Note: you may need to restart the kernel to use updated packages.


In [12]:
import geocoder # import geocoder

In [15]:
# initialize your variable to None
lat_lng_coords = None

postal_code = 'M3A'

# loop until you get the coordinates
while(lat_lng_coords is None):
    g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
    lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]

In [16]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [17]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", data.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


In [18]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [19]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [20]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [21]:
combined_data.shape

(103, 5)

Answer to Question 2: We get 103 rows as expected when we do a inner join, so our data is good.

# Question 3

### Clustering Neighborhoods in Toronto based on the similarities of the venues categories using K-means clustering and Foursquare API.

In [22]:
from geopy.geocoders import Nominatim

In [23]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


In [25]:
pip install folium

     |████████████████████████████████| 94 kB 4.8 MB/s  eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [26]:
import folium

In [49]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

In [54]:
CLIENT_ID = 'L5OJL1L53KTVRL2YFRYTEIWENMWBSB5CE1JYE52H0B2BUUT3' 
CLIENT_SECRET = '52JPH40FH4QPRHFQE40ZTJTRWORCZK1SYIXRTSG2QTQGICF5'
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: L5OJL1L53KTVRL2YFRYTEIWENMWBSB5CE1JYE52H0B2BUUT3
CLIENT_SECRET:52JPH40FH4QPRHFQE40ZTJTRWORCZK1SYIXRTSG2QTQGICF5


In [60]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

In [61]:
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [62]:
venues_in_toronto.shape

(1326, 5)

In [64]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2         Parkwoods               43.753259               -79.329656   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1         Brookbanks Pool                   Pool  
2           Variety Store      Food & Drink Shop  
3  Victoria Village Arena           Hockey Arena  
4               Portugril  Portuguese Restaurant

In [65]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                             Parkwoods   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1311  Mimico NW, The Queensway West, South of Bloor,...   
1312  Mimico NW, The Queensway West, South of Bloor,...   
1313  Mimico NW, The Queensway West, South of Bloor,...   
1314  Mimico NW, The Queensway West, South of Bloor,...   
1315  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                  43.753259               -79.329656         Brookbanks Park   
1                  43.753259               -79.329656         Brookbanks Pool   
2                  43.753259               -79.329656           Variety Store   
3                  43.725882               -79.315572  Victoria Village Arena   
4                  43.725882               -79.315572               Portugril   
...                      ...                      ...                     ...   
1311               43.628841               -79.520999        South St. Burger   
1312               43.628841               -79.520999              Wingporium   
1313               43.628841               -79.520999               Dollarama   
1314               43.628841               -79.520999          Healthy Planet   
1315               43.628841               -79.520999                7-Eleven   

             Venue Category  
0                      Park  
1                      Pool  
2         Food & Drink Shop  
3              Hockey Arena  
4     Portuguese Restaurant  
...                     ...  
1311           Burger Joint  
1312            Wings Joint  
1313         Discount Store  
1314        Supplement Shop  
1315      Convenience Store  

[416 rows x 5 columns]

In [66]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Accessories Store                    Lawrence Manor, Lawrence Heights   
Adult Boutique                                   Church and Wellesley   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Warehouse Store                                      Thorncliffe Park   
Wine Bar                                     Little Portugal, Trinity   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                        Lawrence Manor, Lawrence Heights   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.718518               -79.464763   
Adult Boutique                   43.665860               -79.383160   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Gate                     43.628947               -79.394420   
...                                    ...                      ...   
Warehouse Store                  43.705369               -79.349372   
Wine Bar                         43.653206               -79.400049   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.718518               -79.453512   
Yoga Studio                      43.715383               -79.340923   

                                              Venue  
Venue Category                                       
Accessories Store               Ardene Shoes Outlet  
Adult Boutique                            Seduction  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Gate                                 Gate 8  
...                                             ...  
Warehouse Store                              Costco  
Wine Bar                            Paris Paris Bar  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[234 rows x 4 columns]

In [67]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Adult Boutique  Airport  Airport Food Court  \
0                     0               0        0                   0   
1                     0               0        0                   0   
2                     0               0        0                   0   
3                     0               0        0                   0   
4                     0               0        0                   0   
...                 ...             ...      ...                 ...   
1321                  0               0        0                   0   
1322                  0               0        0                   0   
1323                  0               0        0                   0   
1324                  0               0        0                   0   
1325                  0               0        0                   0   

      Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0                0               0                0                 0   
1                0               0                0                 0   
2                0               0                0                 0   
3                0               0                0                 0   
4                0               0                0                 0   
...            ...             ...              ...               ...   
1321             0               0                0                 0   
1322             0               0                0                 0   
1323             0               0                0                 0   
1324             0               0                0                 0   
1325             0               0                0                 0   

      American Restaurant  Antique Shop  ...  Truck Stop  \
0                       0             0  ...           0   
1                       0             0  ...           0   
2                       0             0  ...           0   
3                       0             0  ...           0   
4                       0             0  ...           0   
...                   ...           ...  ...         ...   
1321                    0             0  ...           0   
1322                    0             0  ...           0   
1323                    0             0  ...           0   
1324                    0             0  ...           0   
1325                    0             0  ...           0   

      Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                                 0                 0            0   
1                                 0                 0            0   
2                                 0                 0            0   
3                                 0                 0            0   
4                                 0                 0            0   
...                             ...               ...          ...   
1321                              0                 0            0   
1322                              0                 0            0   
1323                              0                 0            0   
1324                              0                 0            0   
1325                              0                 0            0   

      Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                         0                0         0            0   
1                         0                0         0            0   
2                         0                0         0            0   
3                         0                0         0            0   
4                         0                0         0            0   
...                     ...              ...       ...          ...   
1321                      0                0         0            0   
1322                      0                0         0            0   
1323                      0                0         0         

In [68]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  Adult Boutique  Airport  \
0         Parkwoods                  0               0        0   
1         Parkwoods                  0               0        0   
2         Parkwoods                  0               0        0   
3  Victoria Village                  0               0        0   
4  Victoria Village                  0               0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Truck Stop  \
0                 0                    0  ...           0   
1                 0                    0  ...           0   
2                 0                    0  ...           0   
3                 0                    0  ...           0   
4                 0                    0  ...           0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 235 columns]

In [69]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Adult Boutique  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0             0.0      0.0                 0.0           0.0             0.0   
1             0.0      0.0                 0.0           0.0             0.0   
2             0.0      0.0                 0.0           0.0             0.0   
3             0.0      0.0                 0.0           0.0             0.0   
4             0.0      0.0                 0.0           0.0             0.0   

   Airport Service  Airport Terminal  American Restaurant  ...  Truck Stop  \
0              0.0               0.0             0.000000  ...         0.0   
1              0.0               0.0             0.000000  ...         0.0   
2              0.0               0.0             0.000000  ...         0.0   
3              0.0               0.0             0.000000  ...         0.0   
4              0.0               0.0             0.045455  ...         0.0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                            0.0               0.0          0.0   
1                            0.0               0.0          0.0   
2                            0.0               0.0          0.0   
3                            0.0               0.0          0.0   
4                            0.0               0.0          0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                    0.0              0.0       0.0          0.0   
1                    0.0              0.0       0.0          0.0   
2                    0.0              0.0       0.0          0.0   
3                    0.0              0.0       0.0          0.0   
4                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 235 columns]

In [70]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [71]:
import numpy as np

In [72]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood      1st Most Common Venue  \
0                                        Agincourt  Latin American Restaurant   
1                           Alderwood, Long Branch                Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                       Bank   
3                                  Bayview Village                       Bank   
4                Bedford Park, Lawrence Manor East                Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0        Clothing Store          Skating Rink        Breakfast Spot   
1              Pharmacy          Skating Rink                   Pub   
2           Coffee Shop                 Diner      Sushi Restaurant   
3    Chinese Restaurant   Japanese Restaurant                  Café   
4        Sandwich Place    Italian Restaurant      Greek Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                Lounge           Yoga Studio      Department Store   
1        Sandwich Place           Coffee Shop                   Gym   
2        Ice Cream Shop           Supermarket          Intersection   
3           Yoga Studio      Department Store     Electronics Store   
4     Indian Restaurant      Sushi Restaurant  Fast Food Restaurant   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0  Eastern European Restaurant             Drugstore             Donut Shop  
1                  Gas Station         Garden Center         Discount Store  
2    Middle Eastern Restaurant         Deli / Bodega      Mobile Phone Shop  
3  Eastern European Restaurant             Drugstore             Donut Shop  
4                      Butcher                  Café                    Pub

In [73]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [74]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

In [75]:
kmeans.labels_[0:100]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 0, 1, 3, 1, 1, 1, 1, 1, 2, 1, 1, 0, 1, 1, 1, 0,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 4, 1,
       3, 1, 1, 1, 1, 1, 1, 3], dtype=int32)

In [76]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [77]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             0.0                  Park   
1  43.725882 -79.315572             1.0           Pizza Place   
2  43.654260 -79.360636             1.0           Coffee Shop   
3  43.718518 -79.464763             1.0        Clothing Store   
4  43.662301 -79.389494             1.0           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue   4th Most Common Venue  \
0                  Pool     Food & Drink Shop             Yoga Studio   
1           Coffee Shop          Intersection            Hockey Arena   
2                  Park                Bakery                    Café   
3     Accessories Store   Arts & Crafts Store  Furniture / Home Store   
4      Sushi Restaurant                  Park                     Bar   

   5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
0           Dance Studio  Eastern European Restaurant             Drugstore   
1  Portuguese Restaurant                        Diner      Department Store   
2         Breakfast Spot                      Theater  Gym / Fitness Center   
3            Event Space           Miscellaneous Shop           Coffee Shop   
4               Beer Bar                Smoothie Shop        Sandwich Place   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Donut Shop               Dog Run    Distribution Center  
1          Dessert Shop    Dim Sum Restaurant    Distribution Center  
2         Historic Site     French Restaurant         Farmers Market  
3              Boutique         Women's Store  Vietnamese Restaurant  
4          Burger Joint         Burrito Place             Restaurant

In [78]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

In [79]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [80]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

In [81]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York             0.0                  Park   
21              York             0.0                  Park   
32       Scarborough             0.0            Playground   
35         East York             0.0                  Park   
85       Scarborough             0.0            Playground   
91  Downtown Toronto             0.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
0                   Pool     Food & Drink Shop                  Yoga Studio   
21         Women's Store                  Pool                 Dance Studio   
32           Yoga Studio           Curling Ice  Eastern European Restaurant   
35     Convenience Store           Yoga Studio                Deli / Bodega   
85                  Park          Intersection                  Yoga Studio   
91            Playground                 Trail                  Yoga Studio   

          5th Most Common Venue        6th Most Common Venue  \
0                  Dance Studio  Eastern European Restaurant   
21  Eastern European Restaurant                    Drugstore   
32                    Drugstore                   Donut Shop   
35            Electronics Store  Eastern European Restaurant   
85                 Dance Studio  Eastern European Restaurant   
91                  Curling Ice                    Drugstore   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0              Drugstore            Donut Shop               Dog Run   
21            Donut Shop               Dog Run   Distribution Center   
32               Dog Run   Distribution Center        Discount Store   
35             Drugstore            Donut Shop               Dog Run   
85             Drugstore            Donut Shop               Dog Run   
91            Donut Shop               Dog Run   Distribution Center   

   10th Most Common Venue  
0     Distribution Center  
21         Discount Store  
32                  Diner  
35    Distribution Center  
85    Distribution Center  
91         Discount Store

In [82]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1          North York             1.0           Pizza Place   
2    Downtown Toronto             1.0           Coffee Shop   
3          North York             1.0        Clothing Store   
4    Downtown Toronto             1.0           Coffee Shop   
6         Scarborough             1.0  Fast Food Restaurant   
..                ...             ...                   ...   
97   Downtown Toronto             1.0           Coffee Shop   
98          Etobicoke             1.0                 River   
99   Downtown Toronto             1.0      Sushi Restaurant   
100      East Toronto             1.0    Light Rail Station   
102         Etobicoke             1.0        Hardware Store   

    2nd Most Common Venue 3rd Most Common Venue   4th Most Common Venue  \
1             Coffee Shop          Intersection            Hockey Arena   
2                    Park                Bakery                    Café   
3       Accessories Store   Arts & Crafts Store  Furniture / Home Store   
4        Sushi Restaurant                  Park                     Bar   
6             Yoga Studio         Deli / Bodega       Electronics Store   
..                    ...                   ...                     ...   
97                   Café            Restaurant      Seafood Restaurant   
98                   Park            Smoke Shop                    Pool   
99              Juice Bar   Martial Arts School        Ramen Restaurant   
100            Skate Park                Garden           Auto Workshop   
102         Burrito Place           Wings Joint         Supplement Shop   

           5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
1          Portuguese Restaurant                 Diner      Department Store   
2                 Breakfast Spot               Theater  Gym / Fitness Center   
3                    Event Space    Miscellaneous Shop           Coffee Shop   
4                       Beer Bar         Smoothie Shop        Sandwich Place   
6    Eastern European Restaurant             Drugstore            Donut Shop   
..                           ...                   ...                   ...   
97                   Art Gallery            Steakhouse                 Hotel   
98              Cuban Restaurant            Donut Shop               Dog Run   
99                Ice Cream Shop     Indian Restaurant  Ethiopian Restaurant   
100         Fast Food Restaurant        Farmers Market         Burrito Place   
102                Tanning Salon         Grocery Store                   Gym   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
1            Dessert Shop    Dim Sum Restaurant    Distribution Center  
2           Historic Site     French Restaurant         Farmers Market  
3                Boutique         Women's Store  Vietnamese Restaurant  
4            Burger Joint         Burrito Place             Restaurant  
6                 Dog Run   Distribution Center         Discount Store  
..                    ...                   ...                    ...  
97                 Bakery             Speakeasy               Tea Room  
98    Distribution Center        Discount Store                  Diner  
99            Escape Room          Dessert Shop               Beer Bar  
100            Comic Shop            Restaurant          Garden Center  
102        Sandwich Place        Discount Store      Convenience Store  

[88 rows x 12 columns]

In [83]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
57   North York             2.0        Baseball Field           Yoga Studio   
101   Etobicoke             2.0                  Pool        Baseball Field   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
57            Event Space           Escape Room     Electronics Store   
101           Yoga Studio         Deli / Bodega     Electronics Store   

           6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
57   Eastern European Restaurant             Drugstore            Donut Shop   
101  Eastern European Restaurant             Drugstore            Donut Shop   

    9th Most Common Venue 10th Most Common Venue  
57                Dog Run    Distribution Center  
101               Dog Run    Distribution Center

In [84]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
64             York             3.0                  Park   
66       North York             3.0                  Park   
83  Central Toronto             3.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
64           Yoga Studio          Dance Studio     Electronics Store   
66     Convenience Store           Yoga Studio         Deli / Bodega   
83           Yoga Studio          Dance Studio     Electronics Store   

          5th Most Common Venue        6th Most Common Venue  \
64  Eastern European Restaurant                    Drugstore   
66            Electronics Store  Eastern European Restaurant   
83  Eastern European Restaurant                    Drugstore   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
64            Donut Shop               Dog Run   Distribution Center   
66             Drugstore            Donut Shop               Dog Run   
83            Donut Shop               Dog Run   Distribution Center   

   10th Most Common Venue  
64         Discount Store  
66    Distribution Center  
83         Discount Store

In [85]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
11  Etobicoke             4.0                Bakery           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
11           Event Space           Escape Room     Electronics Store   

          6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11  Eastern European Restaurant             Drugstore            Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
11               Dog Run    Distribution Center

Answer to Question 3: We have successfully cluster Toronto neighbourhood based on venue categories!